### ISSUE TITLE
Supports dark mode with color-scheme and prefers-color-scheme

### ISSUE THREAD

**Issue by automactic** [openzim/mwoffliner#1136]
Zim files are essentially bundles of html content, to support dark mode, one approach is to use [CSS Color Schemes](https://www.w3.org/TR/css-color-adjust-1/). While still in draft, it's already [widely adopted](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-color-scheme). And I think this more standardized way of handling color scheme is better than rolling our own solution.

**Step 1**: let the browser know the page supports dark and light scheme:
```
:root {
  color-scheme: light dark;
}
```
or 
```
<meta name="color-scheme" content="light dark">
```

**Step 2**: adjust existing css for specific color scheme
```
@media (prefers-color-scheme: dark) {
    xxxx
}
```
I know this step will take up a lot of time and energy, but in most of the situations, the user agent style sheet handles things pretty well.

**Comment by kelson42** [comment]
I really wonder why this is not handle properly like this upstream on Wikipedia.org.

**Comment by Jaifroid** [comment]
Isn't there a proposal for an API in ZIMs that might handle this kind of transform? Currently the different readers handle dark mode by injecting CSS stylesheets. By using an inversion stylesheet and re-inverting certain image type, it's fairly easy to come up with a "universal" (though not perfect) dark mode, and that could be the API. (It's what most of the current readers do.)

Otherwise handling dark mode as a set of themes is something that each Wiki would have to provide, as they would need to be tailored to the different styles used.

**Comment by Lucent** [comment]
Related #2086

**Comment by kelson42** [comment]
After our discussion at https://github.com/openzim/overview/issues/53, we should really implement this.

@benoit74 Do you think it would be easy using approach suggested by @Lucent in #2086?

**Comment by benoit74** [comment]
If I get it correctly, what @Lucent showed in #2086 is that the foundation decided to not resort on standard `prefers-color-scheme` attributes but rather implement their on additional CSS class on `<html>` tag. This means that our current strategy in https://github.com/openzim/overview/issues/53 is not going to easily operate correctly on mwoffliner ZIMs since it means we would have to (automatically) reverse engineer the whole CSS to transform from `.skin-theme-clientpref-night` to `@media (prefers-color-scheme: dark)` ; nothing infeasible, but not easy.

Not speaking about the fact that support for dark mode seems to be well behind the fact that we should be able to ZIM more Mediawikis (support more skins), fetch categories, ... from my PoV.

**Comment by Jaifroid** [comment]
@benoit74 I don't think it's quite so hard. A dev-contributor for the Kiwix JS browser extension contributed a dark mode stylesheet based on Vector (for Wikipedia ZIMs) which is now available in the just released v4.2.0 of the Browser Extension. Granted we have an existing infrastructure in the app for inserting stylesheets for dark themes, but the original was based on inversion, and then reinverting images, which resulted in washed-out colours. The new stylesheet uses "Wikipedia's own dark theme solution" (i.e. the Vector dark CSS), which the contributor simply extracted and downloaded and added into the existing infrastructure in the app. It didn't require editing the HTML of each page, though it does involve injecting the stylesheet into the DOM. Such a solution could be done in JS within the ZIM as part of the long-proposed API.

**_Caveat_**: because it injects the stylesheet into the DOM, there is currently a Flash of Incorrect Theme (FOIT) that is sometimes perceptible between article loads. The way I dealt with this in the PWA (which has had dark theme for years, based on different solutions), is to hide the article (using DOM methods) until the onload event of the stylesheet, and I'm pending adding this or similar fix to the Browser Extension.

**Comment by Jaifroid** [comment]
Despite what I wrote above, and in accordance with https://github.com/openzim/overview/issues/53, I agree that a pure CSS solution is better than using JS to switch between themes (if possible). So mwOffliner should supply both the Vector dark CSS and the Vector light CSS, and ensure these auto-switch with appropriate CSS rules. Plus a way for the reader to detect that ZIM-level support, ideally.

**Comment by Markus-Rost** [comment]
Setting the `skin-theme-clientpref-os` class on the `<html>` element in vector-2022 might already be able to add darkmode for Wikipedia. I'm just not sure if this will create conflicts with wikis who don't have that option enabled. Checking with existing classes on the html tag might help with that.

**Comment by Lucent** [comment]
Using a class for dark mode in conjunction with `prefers-color-scheme` is not an either-or. They can work together. The following pattern uses the OS preference when a cookie is not set, and overrides it when it is. Placed early in the `<head>`, it prevents any flash of unstyled content. The localStorage checks could easily be replaced with checks for the presence of a class injected.

```js
const dark_os_theme = window.matchMedia("(prefers-color-scheme: dark)").matches;
const theme_set = localStorage.getItem("Theme"); // any theme at all set?
const dark_set = localStorage.getItem("Theme") === "Dark";
if (dark_set || (dark_os_theme && !theme_set)) { // if they previously set dark or their OS preference is dark and they have no preference
    document.documentElement.classList.add("Dark");
    document.documentElement.style.colorScheme = "dark";
}
```

**Comment by benoit74** [comment]
> Setting the `skin-theme-clientpref-os` class on the `<html>` element in vector-2022 might already be able to add darkmode for Wikipedia. I'm just not sure if this will create conflicts with wikis who don't have that option enabled. Checking with existing classes on the html tag might help with that.

This is the plan indeed. Unfortunately by tweaking this in existing ZIMs, I came to realize we still miss some important CSS used in this dark mode for ... the infobox. Still didn't had time to assemble details about that, but this is going to be a pain point. 

And as you said, this will work only with vector-2022, and even this is only [a recommendation for wikimedia](https://www.mediawiki.org/wiki/Recommendations_for_night_mode_compatibility_on_Wikimedia_wikis/ha), so not something really strong enough. Some gadgets developers especially for non-wikimedia and other skins uses ... many different techniques.

**Comment by Jaifroid** [comment]
@Lucent That's a JS-based solution, whereas what we are being encouraged to support in readers is a pure CSS solution that looks like:
```
@media (prefers-color-scheme: light) {
   /* light-theme classes */
}

@media (prefers-color-scheme: dark) {
   /* dark-theme classes */
}
```
We're being asked to provide overrides for these colour schemes in the apps, so we can force the colour scheme to light or dark if the user doesn't want to switch their entire OS colour theme. Therefore, we need the solution inside ZIMs to use a standard approach. Otherwise readers will be forced to provide different logic for specific ZIMs.

We also need the ZIM to provide a standard property either in metadata or a property in each HTML page to confirm that it supports at least two CSS colour schemes (light and dark).

**Comment by Markus-Rost** [comment]
> That's a JS-based solution, whereas what we are being encouraged to support in readers is a pure CSS solution that looks like:

There are a multitude of different JS, CSS, and CF worker solutions for themes on wikis. We can't control those at all, though a lot of them do check the media query using JS to change a class or load required modules. That would be the best we can offer.

Some wikis might not check the media query at all and only have a manual JS toggle. We won't be able to support themes for those wikis.

> We're being asked to provide overrides for these colour schemes in the apps, so we can force the colour scheme to light or dark if the user doesn't want to switch their entire OS colour theme. Therefore, we need the solution inside ZIMs to use a standard approach. Otherwise readers will be forced to provide different logic for specific ZIMs.

Just changing the preferred color scheme variable should be fine. The styles actually changing will then depend on how the specific wikis handle the toggle. Be that just CSS or checking with JS.

> We also need the ZIM to provide a standard property either in metadata or a property in each HTML page to confirm that it supports at least two CSS colour schemes (light and dark).

Due to the different solutions mentioned above, we can't reliably tell if a wiki supports themes or not. So we could only take best guesses based on a few known solutions or more likely will need to manually set supported themes as CLI argument. Just always assuming the wiki supports darkmode and the toggle in the reader just not doing anything sometimes might also be a solution.

**Comment by Jaifroid** [comment]
Problem is, it's really bad UX to have a toggle saying "Always use dark theme", but it doesn't work for the great majority of ZIMs. Of course you can have "Always use dark theme if ZIM supports it", but it would be much better for readers to have reliable information about whether the ZIM maker checked and/or included support. Making a ZIM (or a recipe for one) is a fairly conscious activity, so it doesn't seem like a big ask to request that ZIM creators provide a confirmation that their ZIM supports the standard mechanism being developed/agreed in https://github.com/openzim/overview/issues/53. If they don't provide a positive signal, then it can be assumed those mechanisms are unsupported....

**Comment by Markus-Rost** [comment]
> Problem is, it's really bad UX to have a toggle saying "Always use dark theme", but it doesn't work for the great majority of ZIMs.

The CSS media feature is only called `prefers-color-scheme` and calling it a preference to use if available seems like a reasonable approach. That's how it works in normal browsers as well after all.

**Comment by Jaifroid** [comment]
> The CSS media feature is only called `prefers-color-scheme` and calling it a preference to use if available seems like a reasonable approach. That's how it works in normal browsers as well after all.

Still doesn't make it good UX!

It's like having a shower thermostat with settings for "Cold", "Temperature roulette - good luck!", and "Let us decide for you". Users will get confused when clicking "Dark" sometimes works, sometimes doesn't, and sometimes gives them a seizure-inducing strobe effect.

This is a classic anti-pattern: **making user preferences conditional on content support**. When someone chooses "Dark", they mean "I want dark, period, because I work late at night and my eyes hurt" - not "I want dark if the stars align and this particular content happens to support it."

The solution is progressive enhancement: always honour the user's choice, using a great native dark theme when available and graceful fallbacks (like simple CSS inversion, as in the Android app and Kiwix JS) when not. **_For this, we need a reliable way to detect that the ZIM supports a native dark theme._**

"Dark" should actually mean dark, "Auto" should reliably follow OS preferences, and users won't then have to play "guess which ZIM respects my choices today". Otherwise we're training users that their preferences are just suggestions. Cue frustrated/ unhappy/ confused users.. 😕🙃

**Comment by benoit74** [comment]
This issue is about what we will do in mwoffliner, for the general approach the issue is at https://github.com/openzim/overview/issues/53 ; I will put some comments there.

For mwoffliner, the plan is for now two-fold:
- first, enable dark/light preference on wikimedia ZIMs based on vector-2022 (see https://github.com/openzim/mwoffliner/issues/1136#issuecomment-3041798488) and check it works as expected (spoiler: it does not, we miss some CSS for some infobox in dark mode)
- then, find a path for other wikis which support dark/light through a UI switch trigerring JS code setting some classes ; how we will do this is still unsure: either through CSS transformation or some JS, I don't know yet what is the most reliable/feasible solution

**Comment by Jaifroid** [comment]
@benoit74  I agree. All I'm asking for from mwOffliner is a way to signal that an mwOffliner ZIM has been made with explicit dark theme support.

**Comment by Markus-Rost** [comment]
> > Setting the `skin-theme-clientpref-os` class on the `<html>` element in vector-2022 might already be able to add darkmode for Wikipedia. I'm just not sure if this will create conflicts with wikis who don't have that option enabled. Checking with existing classes on the html tag might help with that.
> 
> This is the plan indeed. Unfortunately by tweaking this in existing ZIMs, I came to realize we still miss some important CSS used in this dark mode for ... the infobox. Still didn't had time to assemble details about that, but this is going to be a pain point.

Looks like we are missing the `ext.wikimediamessages.styles` CSS module.

**Comment by Markus-Rost** [comment]
> Looks like we are missing the `ext.wikimediamessages.styles` CSS module.

I have created an upstream issue about the module missing from the API output. Once that is fixed, we should be able to have working dark mode by just changing the class.

https://phabricator.wikimedia.org/T402753

**Comment by benoit74** [comment]
@Jaifroid once #2518 is merged (coming hours or days), mwoffliner will start to honor OS preference of the user / reader when possible, i.e. only for vector 2022 for now.

This will be easily detectable by the fact that `<html>` tag of every articles will have `vector-feature-night-mode-enabled` and `skin-theme-clientpref-os` CSS classes.

We cannot very easily add a metadata at ZIM level since their is no really such a ZIM level decision, but rather a decision article by article to properly set these two CSS classes whenever `vector-feature-night-mode-enabled` and `skin-theme-clientpref-day` or `skin-theme-clientpref-night` classes are returned by the `action=parse` query when retrieving every article. 

I feel like we could hence technically have a ZIM with a mix of articles with support for night mode, and other articles which don't support it.

Is this enough for your usage, or do you feel like a ZIM-level metadata would be much more convenient? We could assume that if main page / first article has the `vector-feature-night-mode-enabled`, then we set the metadata indicating that night mode is supported on this ZIM.

If you need a ZIM-level metadata, what about `X-NightModeEnabled` as name with `true` as value? The `X-` meaning it is not - yet - an openZIM standard metadata, like we already set the `X-ContentDate` on warc2zim ZIMs.

**Comment by Markus-Rost** [comment]
> I feel like we could hence technically have a ZIM with a mix of articles with support for night mode, and other articles which don't support it.

That should only ever happen if the wiki config regarding the night mode got changed while we were scraping the wiki. This would be incredibly rare and not something worth considering here. As such using the main page / first article to check for night mode being enabled should be fine to do.

**Comment by Jaifroid** [comment]
@benoit74 Thanks for the info. It's good news we're getting this ZIM-level support for a propper dark theme! To answer your question, I _could_ if necessary get the info I need from the html of the landing page, though it is quite "late* in the process of displaying an article to be setting app state based on the value found.

However, my main concern is that, while this will work currently for the JS apps that have low-level access to the raw HTML of a page before it is injected into the iframe or webview, that will be harder if we fully activate use of the libzim WASM for ZIM reading, and won't be the case for most other Kiwix apps, I believe.

SO, for the sake of other apps, and also to future-proof the JS apps, I personally think it would be a better solution to have metadata that can be read when loading a ZIM and before looking for the landing page.

Ultimately, you are all in a better position than I am to evaluate the likely way the other apps might implement user choice. At a minimum, I need to ensure not to invert the iframe if user wants dark and ZIM supports dark naively (or if auto theme switching is set and the OS theme is dark and the ZIM has native dark support).

**Comment by Jaifroid** [comment]
Regarding the name of any metadata key (should it be decided to include it), I think we need to have a slightly more precise and future-proof name than `X-NightModeEnabled` (doesn't tell us what colour theme we are dealing with, assumes people only use dark at night...).

How about following the CSS [`color-scheme` spec](https://developer.mozilla.org/en-US/docs/Web/CSS/color-scheme) more closely and using `x-color-scheme` with values being a space-separated string (like in the spec) '`light dark`' (if ZIM supports both). This would allow future values also of '`only light`' or '`only dark`' (as in [spec examples](https://developer.mozilla.org/en-US/docs/Web/CSS/color-scheme#examples)) if a particular ZIM only supports one or the other.

**Comment by benoit74** [comment]
Let me rephrase this suggestions to be sure we are all on the same page.

Scraper would set the `X-ColorScheme` metadata to `light dark` on ZIMs which support both color schemes (ones based on vector-2022 for now, when `vector-feature-night-mode-enabled` is set on main page / first article, or based on mediawiki config value if we achieve to grab it).

Readers would then use this metadata to know that the ZIM supports both colors, and let user decide which color she want to use (probably with a default value being the first one I imagine).

Once user has selected a color, the reader will manipulate the color requested by its user-agent / browser / web component. Which will in turn trigger proper CSS rules which will apply color requested by the user, but also update other "things" of the User-Agent (scrollbars, ...).

And for ZIMs where we support only light (typically), we will set the metadata to `only light` and reader will act accordingly. Or we could even technically set the metadata to `only dark` if we have a ZIM which want only dark (so that for instance reader them change to dark as well).

Is this correct or do I miss something?

**Comment by Jaifroid** [comment]
@benoit74 That's a very well expressed summary, and yes, it's what I was suggesting.

I think the reader would generally interpret absence of the key to be equivalent to `X-ColorScheme: light`, although we do already have some ZIMs that are dark-themed by default, or already follow browser settings. (From memory I think the Mozilla Developer ZIM might be like that, and one or two others - though they are out of scope of mwOffliner.)

**Comment by kelson42** [comment]
@Jaifroid @benoit74 At thi stage: It is not foreseen that readers handle anything optional in the ZIM themselves related to theme. Either dark/light theming works... or does not (and then this is a bug in the ZIM). Please don't build anything in ZIM files which requires changes at reader level.

**Comment by benoit74** [comment]
We are not speaking about **requiring** reader changes.

We are speaking about providing information to readers **wanting** to use it (or needing, like Kiwix JS or PWA).

Otherwise, what do you propose to:
- inform readers already creating dark mode support on their own that they must not do so on some ZIMs because it would conflict with ZIM built-in support of dark mode? (the double inversion, once by the ZIM CSS and once by the reader)
- inform readers which colors are supported in the ZIM so that they can display a nice UI button / combobox / ... to easily switch? or propose to switch to dark during night and stay light during the day? or anything else a reader might want to do with color schemes.

This info should be provided early to the reader, hence the proposition to add a ZIM metadata.

**Comment by kelson42** [comment]
It is not the role of the reader to implement anything of that kind (anymore). This can not work well and we should not do it.

The writer of the scraper is in a far better position (even if not perfect). Therefore there is no information to provide to the reader via dedicated metadata.

Kiwix theming related options are clear and fix: System, Light, Dark. Nothing else, nothing more. We do that and once implemented properly, if there is a strong will to do more, then we will see.

**Comment by Jaifroid** [comment]
@kelson42 I respect your perspective on this, but kindly explain how exactly the ZIM scraper can provide a UI for the user to select the dark or light modes they want? There is no such UI in Wikipedia ZIMs. Providing one seems like a whole different ballgame from what we are discussing. Some users want specific apps to be in light theme while their system is in dark, or vice-versa. It's terrible UX not to provide the option to set the preference at app level IMHO. Be prepared for a deluge of user complaints if they have to set their system theme just to change the app theme.

**Comment by kelson42** [comment]
@Jaifroid No theme changing UI in our ZIM. Please look how this is done on ios and android. It alrady works fine there (at least in dev version).

**Comment by Jaifroid** [comment]
Thanks, but *how* does it work? I don't have access to the dev version of Android or iOS apps, unfortunately. I'd like to understand how the user sets their preference for light or dark.

**Comment by Markus-Rost** [comment]
> [@Jaifroid](https://github.com/Jaifroid) No theme changing UI in our ZIM. Please look how this is done on ios and android. It alrady works fine there (at least in dev version).

Only option I can find in the android app is an automatic invert, which would likely break on ZIMs supporting dark mode natively and is exactly the option @Jaifroid would want the ZIM metadata for to avoid inverting an already dark theme.

**Comment by kelson42** [comment]
> Thanks, but *how* does it work? I don't have access to the dev version of Android or iOS apps, unfortunately. I'd like to understand how the user sets their preference for light or dark.

I can not phrase it better than for 50mn:

"Kiwix theming related options are clear and fix: System, Light, Dark. Nothing else, nothing more."

Duty separation is as follow:
* ZIM provides ligh/dark CSS based themes
* Kiwix provides just the setting to allow user to choose which ZIM based theme to use.

These are the requirements, nothing less or more.

**Comment by Markus-Rost** [comment]
So if the ZIM can't provide light/dark CSS based themes, the Dark option in the reader should just do nothing and leave the content in a light theme?

**Comment by kelson42** [comment]
> So if the ZIM can't provide light/dark CSS based themes, the Dark option in the reader should just do nothing and leave the content in a light theme?

Yes and this is a "bug" in the ZIM.

**Comment by benoit74** [comment]
I'm sorry, but I still don't get how ZIM readers will be able to know that a ZIM is able to provide dark/light or only light or only dark. Letting the reader present an option to go dark (for instance) and then only have reader UI being dark while all the ZIM content is light is not going to look very good. And I don't see how all ZIMs could support dark/light in a near future. Even only for mwoffliner we are saying it is a mostly impossible task to cover all situations outside WMF.

**Comment by kelson42** [comment]
> I'm sorry, but I still don't get how ZIM readers will be able to know that a ZIM is able to provide dark/light or only light or only dark. Letting the reader present an option to go dark (for instance) and then only have reader UI being dark while all the ZIM content is light is not going to look very good.

This is exaxtcty what happens with your browser when Web site has no dark mode.

> And I don't see how all ZIMs could support dark/light in a near future. Even only for mwoffliner we are saying it is a mostly impossible task to cover all situations outside WMF.

Two possible scenarios:
* It is possible technicaly, then the problem is only priorisation (not worry about that)
* If not possible technically, I don't think readers will be in any manner in a better position to solve the problem anyway. So why shoukd we do that?

**Comment by Jaifroid** [comment]
So, we're going round in circles on a discussion we first had in https://github.com/openzim/overview/issues/53 and which wasn't resolved there. As @rgaudin put it there: "It's fundamental to know whether users will have the ability to override the System mode/theme, where and how" ([link](https://github.com/openzim/overview/issues/53#issuecomment-2984619676)).

This is the core of the problem. I can't apply the dark theme in an mwOffliner ZIM (which needs specific CSS classes added or removed AFAIK) unless I know that the ZIM supports it and how to do it. So, given the three theme options I am supposed to provide:

* Light
* Dark  
* System

The only thing I can do with an mwOffliner ZIM without direct CSS manipulation is to set the value to System. To be clear, **there is no standards-provided way for a Web App to override the `prefers-color-scheme` setting provided by the OS** without a lot of JS and injection of CSS variables ([Stack Overflow discussion](https://stackoverflow.com/questions/56300132/how-to-override-css-prefers-color-scheme-setting)). I need to manipulate the specific CSS that mwOffliner provides for this purpose. I *can* manipulate the CSS via DOM methods, and that's fine **so long as I know I need to do it because a user wants to force dark** (or light) **and I know how the specific ZIM implements theming**.

I suspect other reader developers will be in the same position. I understand that mwOffliner can only support what the underlying content sources provide (Wikipedia's approach will differ from Stack Exchange, TED Talks, etc.).

If there is no desire to provide ZIM-level metadata at this point in time, and given these real-world constraints, would it be possible to **document centrally the current approach for theme manipulation** in existing ZIM types? For example:

- How to detect if a ZIM supports dark/light theme switching
- What CSS classes or methods need to be added or removed for a ZIM type
- Any known differences for other content types

This documentation would help reader devs provide consistent theme experiences to users.

Don't get me wrong, I hugely appreciate the work done with mwOffliner (kudos to @Markus-Rost and @benoit74 for their hard work on it), and it's a system I'm **very keen** to support. I'm just trying to understand how to implement proper theme support in practical terms.

**Comment by Jaifroid** [comment]
I note the latest ZIMs now have dark theme seemingly turned on by default if the OS is set to dark. This is the case of the latest golf ZIM (https://browse.library.kiwix.org/viewer#wikipedia_en_golf_maxi_2025-10). My OS is set to dark, so I'm seeing the below in the JS readers and in library.kiwix.org.

So again I ask, what is the agreed-upon method for the reader to detect that the ZIM supports this dark theme? How else can I know to switch the app's theme? @kelson42 You and I had a meeting about this, and you agreed to think about a solution. In the meantime, ZIMs are released that automatically switch to dark without a documented method (unless I've missed it). I now class this as a UI bug/regression, and it's not great that despite all the discussions, I now have to scramble to "fix" this. 🙁

<img width="1317" height="1163" alt="Image" src="https://github.com/user-attachments/assets/8511167a-8cb4-4e06-bb75-bf89625f4ac4" />

**Comment by Jaifroid** [comment]
So, I've hacked a detection algorithm for Wikipedia/Wikimedia ZIMs which I've incorporated into the PWA (and I plan to incorporate into Kiwix JS BE asap), but I'm sure you can all see the practical problems with a mechanism like this.
```javascript
var usesProvidedWikipediaTheme = false;
var htmlElement = doc.documentElement;
if (htmlElement) {
    // First check if we've already detected and marked this article
    if (htmlElement.hasAttribute('data-kiwix-has-skin-theme-clientpref')) {
        usesProvidedWikipediaTheme = htmlElement.getAttribute('data-kiwix-has-skin-theme-clientpref') === 'true';
    } else if (htmlElement.classList) {
        // Initial detection: check if the html element has one of the theme preference classes
        usesProvidedWikipediaTheme = htmlElement.classList.contains('skin-theme-clientpref-night') ||
                                htmlElement.classList.contains('skin-theme-clientpref-os');
        // Remember this detection for subsequent theme switches on the same article (since we need to remove the class to override)
        htmlElement.setAttribute('data-kiwix-has-skin-theme-clientpref', usesProvidedWikipediaTheme ? 'true' : 'false');
    }
}
console.log('[Theme Detection] Article uses Wikipedia\'s auto theme system:', usesProvidedWikipediaTheme);
```
In case it's not obvious, this is highly specific to Wikimedia ZIMs, it's fragile and could easily break with a change of stylesheet. We desperately need a more generic detection method, ideally a metadata entry, and even better a JS API to override the switching mechanisms for each ZIM in a generic way that abstracts the specific details of how each ZIM's content implements dark and light themes.

Am I the only app developer experiencing an issue here? How are other ZIM readers planning to manipulate the display of these themes? It's really presumptive, let alone terrible UX, to force users to switch their global OS setting as the only mechanism for changing the theme. If that weren't the case, why do so many web sites provide in-page light/dark switching icons nowadays?

**Comment by benoit74** [comment]
I can only say that I second @Jaifroid frustration, feedback, suggestion, recommendations. I gave up on convincing @kelson42 his expectations on UX might not be the ones of most users, I've already lost too much time explaining things in this issue without being heard at all. But I'm ready to help, and still frustrated as well.

Btw, your detection algorithm is indeed correct, this is the only support for "dark" theme introduced in mwoffliner for now. It does cover a bit more than Wikimedia, mostly everything using vector-2022 and following [Mediawiki recommendations for wikimedia wikis](https://www.mediawiki.org/wiki/Recommendations_for_night_mode_compatibility_on_Wikimedia_wikis/ha) (which we could expect some users do)

**Comment by kelson42** [comment]
> Am I the only app developer experiencing an issue here? How are other ZIM readers planning to manipulate the display of these themes? It's really presumptive, let alone terrible UX, to force users to switch their global OS setting as the only mechanism for changing the theme. If that weren't the case, why do so many web sites provide in-page light/dark switching icons nowadays?

We had talk a few weeks ago and your situation is unique in Kiwix reader portfolio: you are the only one having an SDK which does not allow to make a switch at the reader level.

For the moment unsure about the appropriate answer to this problem.

**Comment by Jaifroid** [comment]
> We had talk a few weeks ago and your situation is unique in Kiwix reader portfolio: you are the only one having an SDK which does not allow to make a switch at the reader level.

There is indeed no API in JS that overrides the OS-level preferences, according to my research (several sites say it can't be done by design). I think the Kiwix Library also has this issue, albeit its UI/Chrome is rather minimal. The latest nightly of Kiwix Desktop I could find for Windows (last built 16-10, Linux one is currently missing a library to run in WSL strangely) doesn't have any controls for switching the UI, but the golf_10-25 ZIM is displayed with light colour scheme on my dark-mode system, suggesting that OS-level environment is not communicated to the app currently (or it is not acted upon).

The appropriate answer to the problem from my perspective is to provide a universal detection method (a simple flag), and (in due course) an API for overriding/switching in the ZIM's HTML, which was the original intention since #596. It's a shame that was dropped IMHO, as it would have been an elegant solution.

While we are just talking about Wikimedia ZIMs, I can support those with the kind of hack above. But you can see the issue if auto-theme switching were to become the norm with myriad different CSS classes across different ZIM flavours, as I (or a future dev) would have to maintain a whole library of methods to do the detection and override.

I know Kiwix JS is important to the org (latest metrics on just those delivered by stores show many thousands of people using them) and you have provided strong support for us in the past, so I appreciate any software infrastructural support that these apps (at least) need.

**Comment by Markus-Rost** [comment]
Oh, I'm only now understanding the issue you have in kiwix-js and that's indeed nearly impossible to fix.

Even mwoffliner itself doesn't really know if a wiki supports prefers-color-scheme or not. And the classes needed to toggle between set light and dark themes can be even more different between wikis, with some requiring additional css modules etc.

A library of methods to guess and switch on the different wikis would be needed in any solution, either in the reader or mwoffliner directly. However having the API in mwoffliner likely means every page load will have a FOUC, which could be avoided by the reader having the API and modifying the html before it gets rendered by the client.

Regarding Wikimedia ZIMs at least, the solution is fairly simple. Mwoffliner already uses the class detection to do some replacement to enable the automatic theme toggle in the first place. @Jaifroid you could just reverse the replacements again to get your preferred theme.

https://github.com/openzim/mwoffliner/blob/0fd2ab27c0644826a3a90d9664bcf8c358fdd199/src/util/articles.ts#L70-L72

**Comment by rgaudin** [comment]
Haven't been much involved in this discussion so I'll try to summarize what I understand and my opinion (I think I'm leaning toward @kelson42's).

- Light/Dark theming is an OS feature
- It's assumed that users don't always have control over this OS feature (permission, ease of access, etc)
- Web browsers, understanding previous point, offer an option to override it (the Keep System, Light, Dark options)
- OS vendors usually discourage overriding this at application level but don't prevent it because there are valid use cases (like web browsers on which people spend their day, reading)
- Ideally, websites wanting to adjust to themes should use the standard mechanism (as we did in that test ZIM)

I think above points are shared. Now, as to how Kiwix should embrace Dark mode, here's what's been decided (to the best of my knowledge):

- We will (slowly, over time) add dark mode support to our ZIMs
  - using the standard approach demoed in the [light/dark test ZIM](https://dev.library.kiwix.org/#lang=&q=light) for those we have control of the UI
  - enable/tweak/fix for scrapers where source website support it (need to use the std approach): mwoffliner.
  - hope for the best regarding zimit-made ZIMs
- Readers will have their chrome support dark mode
- Readers, similarly to browsers (because they serve shared purpose) will offer a similar option to override it (Keep System, Light, Dark options)

That's the strategy as I understood it and I am aligned with it.

I believe @benoit74 is worried about the effort/long transition between now where readers start to support Dark (chrome is displayed dark) but content is not (content is bright white) and how that could be considered deceptive. I don't share this conclusion (it's the case every where on the web) and if that's a concern, maybe a separate ticket would better serve this side topic.

I understand @Jaifroid is not able to apply this because in PWA/Extension contexts, there is no way to override the theme and thus those readers would not have the option.

My opinion here is that when in PWA/Extension context, the reader is not a standalone reader but part of the browser and it should be the browser's responsibility to allow overriding the theme. Reader should only apply the same theme mechanism for its own chrome and not offer an option to override it. I understand we could have diverging opinions on this.
For electron-based versions, then I'd expect the reader to have the ability to override it.

Also, and that's linked to the previous point, there's the fact that Kiwix JS readers have long offered a theming option. This option goes beyond what's discussed here because it can apply dark mode to ZIMs that don't support it. I'm not a fan of such generic strategies (we've tested this in the past) but it certainly is comfortable to some and given the large user base, there might be a large number of users accustomed to it.

<img width="436" height="256" alt="Image" src="https://github.com/user-attachments/assets/f21fa3ea-dc0c-4799-88cf-bb332e34ddab" />

If I understand correctly, @Jaifroid wants to keep this option working until the transition of all ZIMs is ~over (or forever if we consider zimit-made ones).

Again, if I understand correctly, this somewhat assumes that all ZIMs are light mode and needs to be able to _detect_ those that _natively_ support dark in order to disable the invert mechanism. I think it was suggested that we add a metadata to ZIM if we know it natively supports dark mode.

On one hand, I find it unfair to put on ZIM creation requirements that satisfy a hack in a single reader for a feature we've deemed not worth implementing everywhere, but on the other hand, the effort is small and we will anyway have to identify (albeit not necessarily via a metadata) which ZIMs are dark-mode ready and which still are not.

I guess there are three possible outcomes:
- **A**. We add a way to identify those ZIMs
- **B**. @Jaifroid drops the feature
- **C**. @Jaifroid keeps the feature and has to constantly battle to keep it somewhat working.

**Comment by Jaifroid** [comment]
@rgaudin Thank you for your careful and thorough summary of the issues.

I want to make sure I understand the proposed approach for readers like Kiwix Desktop: how will they implement UI controls for switching between dark and light modes when they have no way to know how/if the ZIM content will respond to those changes? It's a Catch-22 situation. Even once all Wikimedia ZIMs support OS-level theme changes, there will remain a vast number of ZIMs where content won't automatically respond using `window.matchMedia('(prefers-color-scheme)')`. Without some detection mechanism, there will be a long transition period - or perhaps permanent state - where app UI and content UI can clash dramatically. This happened immediately with the KJS apps when the first auto-switching ZIMs appeared (see screenshot above), and I've had to scramble to address it. I've released a fix for the PWA and will do the same for the Browser Extension as time permits from my day job.

The core challenge is this: you detect that the OS (or browser) prefers dark mode, so you switch your app UI to dark - but you have no way to know if the content will also switch. This creates genuinely confusing UX. On top of that, I have additional JavaScript API limitations most others don't face.

Regarding PWAs: I respectfully disagree that they should be treated as "just browser apps". Significant engineering has gone into making them installable at OS level with isolated contexts and [native-level APIs](https://whatpwacando.today/), across Chromium, Safari, and Firefox (on Android, but now also installable in Firefox 143+ for Windows). They behave as standalone applications to users. And as you note, Electron apps package their own browser context entirely - even if I wanted to ask users to toggle their entire browser's/OS's theme settings just to read Wikipedia when the sun comes out from behind a cloud (which seems quite presumptuous), I'd still need app-level controls for Electron.

I also want to clarify the scope we're discussing: the Kiwix JS ecosystem includes browser extensions in four stores (Chrome, Edge, Mozilla, Ubuntu Touch), a PWA, three Microsoft Store apps, Electron apps for all three desktop OSes and architectures, plus dedicated WikiMed and Wikivoyage packages for the same targets. The Microsoft Store apps alone register 20,000 installs over the last 12 months. The Chrome extension store registered 7,000 installs for the same period. I believe the PWA serves significantly more than all these due to its low-friction installs on every desktop and mobile platform, and instant updates, but we don't collect telemetry. As _**the sole remaining regular volunteer developer**_ on these since mossroy departed, I've been working to extend Kiwix's reach to platforms and use cases that complement the other apps. I hope you'll understand why I merely ask for solutions that allow me to continue serving this community effectively.

**That said, I genuinely appreciate the time you've taken to think through these issues systematically.** I'm not suggesting the organization should reorient around the JS apps' architecture (though I do appreciate past accommodations like the inline JavaScript fix, now surpassed). What I am suggesting is that the UX problem I've encountered - mismatched UI themes between chrome and content - may actually be a *general* problem during any transition period, not unique to the JavaScript apps. If that's the case, then solutions like ZIM metadata for dark-mode detection would benefit the entire ecosystem.

The three options you outlined are clear:

A. Add a way to identify dark-mode-ready ZIMs
B. I drop the manual theme toggle feature
C. I maintain the feature through increasingly complex hacks.

**I'd strongly advocate for option A.** A simple metadata flag (e.g., `ColorScheme: light dark` for auto-switching ZIMs, `ColorScheme: light` for fixed-light ZIMs) would be minimal effort to implement and would allow *any* reader to make intelligent decisions about theme handling. This follows the familiar CSS `color-scheme` syntax and would be immediately intuitive to developers. This seems like good infrastructure that could benefit all readers during the transition and beyond.

I am entirely open to other solutions of course, and greatly appreciate your attention to this matter.

**Comment by Jaifroid** [comment]
@Markus-Rost - Thank you for the extremely helpful pointer to the mwOffliner class replacements. I've already implemented a workaround in the PWA for Wikimedia ZIMs by detecting and manipulating classes on the html element. It works well for the Wikimedia case, but I'll look into whether your suggestion would be even simpler (for the Browser Extension work I still need to do). As you note, this detection and override problem would exist across any wiki with different theming implementations. 

A metadata flag at ZIM level would solve the detection problem (knowing which ZIMs natively respond to theme changes) for everyone, but for the JS apps, it seems I'd still need to maintain a library of methods for manipulating different CSS classes across different wiki flavours. This is why a standardized JavaScript API (as originally proposed in #596) would have been the ideal long-term solution - each ZIM creator could simply provide an abstracted, standard interface for theme switching (e.g., `zimTheme.set('light')` or `zimTheme.set('dark')`), eliminating the need for reader-side per-ZIM hacks or centralized method libraries entirely. I understand this approach is unfortunately no longer being pursued, so a metadata flag would at least solve half the problem for most of the apps.

**Comment by Markus-Rost** [comment]
@Jaifroid I see no issue with adding a metadata flag, however for most mwoffliner ZIMs that value would be "unknown". The ZIM might have a light, dark or automatic theme but the scraper has no way of knowing which is the case. Currently only for Wikimedia ZIMs the value could be guessed fairly reliably based on the class.

I think just having the app UI always respect the users prefered color scheme would be the best solution for "unknown" theme ZIMs. This behavior would also match how all browsers behave by default regarding their UI.

I understand the difficulty for JS based readers which are unable to set the value for the prefers-color-scheme media query directly, but I can't think of a good solution to this besides pointing the user to the browser preference.

Regarding a standardized Javascript API for the theme toggle, that would be even less reliable than the ZIM metadata guesses. Wikis which can be guessed in the future might not have a way to toggle the theme besides the media query.
This would also just be the library of methods for class manipulation, just centralized at scraper level instead of the readers. But due to being on the ZIM level, this would then likely cause a FOUC (Flash Of Unstyled Content) on every page load before the correct theme is set by JS. This FOUC could be avoided by having the class manipulation done at reader level, before the client starts rendering the html.

**Comment by kelson42** [comment]
> > Am I the only app developer experiencing an issue here? How are other ZIM readers planning to manipulate the display of these themes? It's really presumptive, let alone terrible UX, to force users to switch their global OS setting as the only mechanism for changing the theme. If that weren't the case, why do so many web sites provide in-page light/dark switching icons nowadays?
> 
> We had talk a few weeks ago and your situation is unique in Kiwix reader portfolio: you are the only one having an SDK which does not allow to make a switch at the reader level.
> 
> For the moment unsure about the appropriate answer to this problem.

I had a bit of doubt this morning, but after these last comments and other live discussion, I'm now convinced that the problem we have been discussing since weeks around the Kiwix JS PWA and extensions is not a problem but a feature: it is not possible (in a clean manner) to change the wanted CSS theme at Kiwix level and this is good like this. This is not the duty of Kiwix, but the duty of the browser.

**Comment by kelson42** [comment]
> > We had talk a few weeks ago and your situation is unique in Kiwix reader portfolio: you are the only one having an SDK which does not allow to make a switch at the reader level.
> 
> There is indeed no API in JS that overrides the OS-level preferences, according to my research (several sites say it can't be done by design). I think the Kiwix Library also has this issue, albeit its UI/Chrome is rather minimal. The latest nightly of Kiwix Desktop I could find for Windows (last built 16-10, Linux one is currently missing a library to run in WSL strangely) doesn't have any controls for switching the UI, but the golf_10-25 ZIM is displayed with light colour scheme on my dark-mode system, suggesting that OS-level environment is not communicated to the app currently (or it is not acted upon).

All of these are reader issues which are tracked in respective issues and will be fixed.

> The appropriate answer to the problem from my perspective is to provide a universal detection method (a simple flag), and (in due course) an API for overriding/switching in the ZIM's HTML, which was the original intention since [#596](https://github.com/openzim/mwoffliner/issues/596). It's a shame that was dropped IMHO, as it would have been an elegant solution.
> 
> While we are just talking about Wikimedia ZIMs, I can support those with the kind of hack above. But you can see the issue if auto-theme switching were to become the norm with myriad different CSS classes across different ZIM flavours, as I (or a future dev) would have to maintain a whole library of methods to do the detection and override.
> 
> I know Kiwix JS is important to the org (latest metrics on just those delivered by stores show many thousands of people using them) and you have provided strong support for us in the past, so I appreciate any software infrastructural support that these apps (at least) need.

This whole stuff is not related to Kiwix JS, it is related to PWA (and extensions). In the context of PWA (and extension) the theming choice is not (and never have been) a duty of Kiwix. The problem is not accepting that fact.

**Comment by kelson42** [comment]
I have created the issue #2564 to discuss how we could extend the dark mode further.

**Comment by Jaifroid** [comment]
> The problem is not accepting that fact.

We clearly have a different understanding of the meaning of the word "fact".

**Comment by Jaifroid** [comment]
> the scraper has no way of knowing which is the case

@Markus-Rost Thanks for the clarification. I was thinking this should be the job of the human who curates a ZIM (whether by creating the recipe or using the software to scrape a site). They should set such a flag. But looks like this is a dead end., sadly.
